File: aoc15/day_05/day_05.ml
let lines =
  List.map
    (fun l -> List.init (String.length l) (String.get l))
    (Input.lines "input")
in

let vowel c = match c with 'a' | 'e' | 'i' | 'o' | 'u' -> true | _ -> false in
let vowels line = List.length (List.filter vowel line) >= 3 in

let rec double line =
  match line with
  | [] | _ :: [] -> false
  | f :: s :: ls -> if f == s then true else double (s :: ls)
in

let rec contains (x, y) line =
  match line with
  | [] | _ :: [] -> false
  | f :: s :: ls -> if f == x && s == y then true else contains (x, y) (s :: ls)
in

let no_forbidden line =
  Bool.not
    (List.exists
       (fun pair -> contains pair line)
       [ ('a', 'b'); ('c', 'd'); ('p', 'q'); ('x', 'y') ])
in

let rec twice_double line =
  match line with
  | [] | _ :: [] -> false
  | f :: s :: ls -> if contains (f, s) ls then true else twice_double (s :: ls)
in

let rec triplet line =
  match line with
  | [] | _ :: [] | [ _; _ ] -> false
  | f :: s :: t :: ls -> if f == t then true else triplet (s :: t :: ls)
in

let is_match rules line = List.for_all (fun r -> r line) rules in

let first =
  List.length (List.filter (is_match [ vowels; double; no_forbidden ]) lines)
in
let second =
  List.length (List.filter (is_match [ twice_double; triplet ]) lines)
in
Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_05/day_05.go
package main

import (
	"aoc/collections"
	"bufio"
	"fmt"
	"os"
	"strings"
)

func threevowels(str string) bool {
	return collections.CountFunc([]rune(str), vowel) >= 3
}

func twice(str string) bool {
	for i := 0; i < len(str); i++ {
		if i > 0 && str[i] == str[i-1] {
			return true
		}
	}
	return false
}

func vowel(c rune) bool {
	return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
}

func badpairs(str string) bool {
	return strings.Contains(str, "ab") || strings.Contains(str, "cd") ||
		strings.Contains(str, "pq") || strings.Contains(str, "xy")
}

func niceFirst(str string) bool {
	return threevowels(str) && twice(str) && !badpairs(str)
}

func twicepair(str string) bool {
	for i := 0; i < len(str); i++ {
		if i >= 2 && strings.Contains(str[i:], string(str[i-2])+string(str[i-1])) {
			return true
		}
	}
	return false
}

func triplet(str string) bool {
	for i := 0; i < len(str); i++ {
		if i >= 2 && str[i] == str[i-2] {
			return true
		}
	}
	return false
}

func niceSecond(str string) bool {
	return twicepair(str) && triplet(str)
}

func main() {
	input, _ := os.Open("input")
	scanner := bufio.NewScanner(input)
	strings := []string{}
	for scanner.Scan() {
		strings = append(strings, scanner.Text())
	}

	fmt.Println("First:", collections.CountFunc(strings, niceFirst))
	fmt.Println("Second:", collections.CountFunc(strings, niceSecond))
}
